source/core/rtc_adapter/AudioSendAdapter.cc

```diff
@@ -0,0 +1,167 @@
+
+namespace rtc_adapter {
+
+AudioSendAdapterImpl::AudioSendAdapterImpl(CallOwner* owner,  const RtcAdapter::Config& config)
+    : m_frameFormat(FRAME_FORMAT_UNKNOWN)
+    , m_lastOriginSeqNo(0)
+    , m_seqNo(0)
+    , m_ssrc(0)
+    , m_ssrc_generator(SsrcGenerator::GetSsrcGenerator())
+    , m_rtpListener(config.rtp_listener)
+    , m_statsListener(config.stats_listener)
+{
+    m_ssrc = m_ssrc_generator->CreateSsrc();
+    m_ssrc_generator->RegisterSsrc(m_ssrc);
+    m_taskRunner.reset(new owt_base::WebRTCTaskRunner("AudioSendAdapterImpl"));
+    m_taskRunner->Start();
+    init();
+}
+
+AudioSendAdapterImpl::~AudioSendAdapterImpl() {
+    close();
+    m_taskRunner->Stop();
+    m_ssrc_generator->ReturnSsrc(m_ssrc);
+    boost::unique_lock<boost::shared_mutex> lock(m_rtpRtcpMutex);
+}
+
+int AudioSendAdapterImpl::onRtcpData(char* data, int len) {
+    boost::shared_lock<boost::shared_mutex> lock(m_rtpRtcpMutex);
+    m_rtpRtcp->IncomingRtcpPacket(reinterpret_cast<uint8_t*>(data), len);
+    return len;
+}
+
+void AudioSendAdapterImpl::onFrame(const Frame& frame) {
+    if (frame.length <= 0)
+        return;
+
+    if (frame.format != m_frameFormat) {
+        m_frameFormat = frame.format;
+        setSendCodec(m_frameFormat);
+    }
+
+    if (frame.additionalInfo.audio.isRtpPacket) {
+        // FIXME: Temporarily use Frame to carry rtp-packets
+        // due to the premature AudioFrameConstructor implementation.
+        updateSeqNo(frame.payload);
+        if (m_rtpListener) {
+            m_rtpListener->onAdapterData(reinterpret_cast<char*>(frame.payload), frame.length);
+        }
+        return;
+    }
+
+    int payloadType = getAudioPltype(frame.format);
+    if (payloadType == INVALID_PT)
+        return;
+
+    boost::shared_lock<boost::shared_mutex> lock(m_rtpRtcpMutex);
+    if (!m_rtpRtcp->OnSendingRtpFrame(frame.timeStamp, -1, payloadType, false)) {
+        RTC_DLOG(LS_WARNING) << "OnSendingRtpFrame return false";
+        return;
+    }
+    const uint32_t rtp_timestamp = frame.timeStamp + m_rtpRtcp->StartTimestamp();
+    // TODO: The frame type information is lost. We treat every frame a kAudioFrameSpeech frame for now.
+    if (!m_senderAudio->SendAudio(webrtc::AudioFrameType::kAudioFrameSpeech, payloadType, rtp_timestamp,
+                                  frame.payload, frame.length)) {
+        RTC_DLOG(LS_ERROR) << "ChannelSend::SendData() failed to send data to RTP/RTCP module";
+    }
+}
+
+bool AudioSendAdapterImpl::init() {
+    m_clock = Clock::GetRealTimeClock();
+    m_eventLog = std::make_unique<webrtc::RtcEventLogNull>();
+
+    RtpRtcp::Configuration configuration;
+    configuration.clock = m_clock;
+    configuration.audio = true;  // Audio.
+    configuration.receiver_only = false;
+    configuration.outgoing_transport = this;
+    configuration.event_log = m_eventLog.get();
+    configuration.local_media_ssrc = m_ssrc;//rtp_config.ssrcs[i];
+
+    m_rtpRtcp = RtpRtcp::Create(configuration);
```

Comment:
so for audio/video send, you have different Rtp/Rtcp instance? why it's not designed to share one?

Reply:
We'll share it later